dataset/bins.cpp

```diff
@@ -405,4 +405,53 @@ Dataset sum(const Dataset &d) {
   return apply_to_items(d, [](auto &&... _) { return buckets::sum(_...); });
 }
 
+Variable mean(const Variable &data) {
+  auto type = variable::variableFactory().elem_dtype(data);
+  type = type == dtype<bool> ? dtype<int64_t> : type;
+  const auto unit = variable::variableFactory().elem_unit(data);
+
+  Variable means;
+  if (variable::variableFactory().hasVariances(data))
+    means = Variable(type, data.dims(), unit, Values{}, Variances{});
+  else
+    means = Variable(type, data.dims(), unit, Values{});
+
+  if (data.dtype() == dtype<bucket<DataArray>>) {
+    const auto &&[indices, dim, buffer] = data.constituents<DataArray>();
+    if (const auto mask_union = irreducible_mask(buffer.masks(), dim);
+        mask_union.is_valid()) {
+      const auto masked = applyMask(buffer, indices, dim, mask_union);
+      variable::sum_impl(means, masked);
+
+      // Trick to get the sizes of bins if masks are present - bin the masks
+      // using the same dimension & indices as the data, and then sum the
+      // inverse of the mask to get the number of unmasked entries.
+      auto sizes = makeVariable<int64_t>(masked.dims());
+      variable::sum_impl(sizes, make_bins(indices, dim, ~mask_union));
```

Comment:
Might want to use https://github.com/scipp/scipp/blob/53b7faad1dbe02653e91936bc6ec1104c5ed9b9d/variable/bins.cpp#L79 here instead, since you are reusing existing indices.